### PR TITLE
Histogram events & bin hover label improvements

### PR DESCRIPTION
**Pull request by alexcjohnson**
Implements #2071 (histogram events get the input points selected) and #2086 (better display of data ranges for histogram bins) for both `histogram` and the 2d histogram types (`histogram2d` and `histogram2dcontour`)

### CODE REVIEW

File: src/lib/search.js [hunk 1]
```diff
@@ -12,6 +12,11 @@
 var isNumeric = require('fast-isnumeric');
 var loggers = require('./loggers');
 
+// don't trust floating point equality - fraction of bin size to call
+// "on the line" and ensure that they go the right way specified by
+// linelow
+var roundingError = 1e-9;
```
Comment (alexcjohnson) [inline src/lib/search.js]: I'm a little surprised that this hadn't come up before (other than the test [below](https://github.com/plotly/plotly.js/pull/2113/commits/6dc99047d171ac7acea56d5471cb0cbdff2ecdb2#diff-8c89958f1286cac36ab2325921671724R426) that was actually wrong before), and obviously it's a bit questionable exactly what small fraction to put here... but if we just use straight `<` vs `<=` etc it's pretty easy to trick `findBin` on the edges.

File: src/components/fx/hover.js [hunk 2]
```diff
@@ -1063,14 +1063,13 @@ function cleanPoint(d, hovermode) {
     d.y0 = Lib.constrain(d.y0, 0, d.ya._length);
     d.y1 = Lib.constrain(d.y1, 0, d.ya._length);
 
-    // and convert the x and y label values into objects
-    // formatted as text, with font info
+    // and convert the x and y label values into formatted text
     if(d.xLabelVal !== undefined) {
-        d.xLabel = ('xLabel' in d) ? d.xLabel : Axes.hoverLabelText(d.xa, d.xLabelVal);
+        d.xLabel = ('xLabel' in d) ? d.xLabel : getDimText(d, 'x');
```
Comment (alexcjohnson) [inline src/components/fx/hover.js]: @etpinard I made `getDimText` in parallel with you adding `d.(x|y)Label` and `Axes.hoverLabelText` - I suppose in principle we could use `d.(x|y)Label` for this too instead of adding `d.(x|y)LabelVal(0|1)`, maybe that would actually be simpler... let me take a look.

File: src/traces/histogram/calc.js [hunk 3]
```diff
@@ -171,15 +171,19 @@ module.exports = function calc(gd, trace) {
...
             };
-            if(uniqueValsPerBin) {
-                cdi.p0 = cdi.p1 = (inputPoints[i].length) ? pos0[inputPoints[i][0]] : pos[i];
-            }
-            else {
-                cdi.p0 = roundFn(binEdges[i]);
-                cdi.p1 = roundFn(binEdges[i + 1], true);
+
+            // pts and p0/p1 don't seem to make much sense for cumulative distributions
+            if(!cumulativeSpec.enabled) {
```
Comment (alexcjohnson) [inline src/traces/histogram/calc.js]: A questionable decision... this gets at some of the same issues that led to `cumulative.currentbin` but in principle a CDF shows the sum of all the data prior to a specific point, not the data within that bin. perhaps I should shift `p0` and `p1` depending on `currentbin` so that the hover label at least gets the description "all the data prior to X" precisely correct, even though it'll be a different value than the center of the bar...

And then what about `pts`? You could say it should be included and again should be "all the data prior to X" but then it would be meaningless to select a single bar and not all the bars before it. Alternatively `pts` could contain "all the data that was added in this bar" which might be more useful in terms of selecting a single bar, but it's not really what that bar means.

File: test/jasmine/tests/histogram_test.js [hunk 4]
```diff
@@ -664,3 +666,207 @@ describe('Test histogram', function() {
...
+        _test(0.05, 0.3, [0, 2, 4, 6], false, [0, 1.9, 2, 3.9, 4, 5.9]);
+        _test(0, 0.1, [0, 1, 2], false, [0, 0.9, 1, 1.9]);
+        _test(0, 0.001, [0, 1, 2], false, [0, 0.999, 1, 1.999]);
+        _test(0.1, 0.1, [115, 125, 135], false, [115, 124.9, 125, 134.9]);
+        _test(0.1, 0.01, [115, 125, 135], false, [115, 124.99, 125, 134.99]);
+        _test(10, 100, [5000, 6000, 7000], false, [5000, 5900, 6000, 6900]);
+
+        // too small of a right gap: stop disambiguating data at the edge
+        _test(0, 0.0009, [0, 1, 2], false, [0, 1, 1, 2]);
+        _test(0.1, 0.009, [115, 125, 135], false, [115, 125, 125, 135]);
```
Comment (alexcjohnson) [inline test/jasmine/tests/histogram_test.js]: @etpinard @chriddyp as discussed offline this morning - compare to lines 699 and 701 above that are just on the other side of the cutoff I implemented in https://github.com/plotly/plotly.js/pull/2113/commits/ad0e08f14bc18e3b8f5a570d5372db15696d86d0 - seem reasonable?

Comment (alexcjohnson) [reply test/jasmine/tests/histogram_test.js]: to be clear, what these tests are indicating is that the hover labels for these bins will be `115 - 124.99` and `125 - 134.99` on the disambiguated side of the cutoff, and `115 - 125` and `125 - 135` on the too-many-digits-to-disambiguate side.

File: test/jasmine/tests/histogram_test.js [hunk 5]
```diff
@@ -664,3 +666,207 @@ describe('Test histogram', function() {
...
+        _test(0, 28 * day, [cn1_00, cn1_01, cn1_02], cn,
+            [cn1_00, Lib.dateTime2ms('2000-12-01', cn), cn1_01, Lib.dateTime2ms('2001-12-01', cn)]);
+        _test(14 * day, 14 * day, [cn1_00 - 14 * day, cn1_01 - 14 * day, cn1_02 - 14 * day], cn,
+            [cn1_00, Lib.dateTime2ms('2000-12-01', cn), cn1_01, Lib.dateTime2ms('2001-12-01', cn)]);
+
+        _test(0, 353 * day, [cn1_00, cn1_10, cn1_20], cn,
+            [cn1_00, Lib.dateTime2ms('2009-01-01', cn), cn1_10, Lib.dateTime2ms('2019-01-01', cn)]);
+        // occasionally we give extra precision for world dates (month when it should be year
+        // or day when it should be month). That's better than doing the opposite... not going
+        // to fix now, too many edge cases, better not to complicate the logic for them all.
```
Comment (alexcjohnson) [inline test/jasmine/tests/histogram_test.js]: The algorithm I came up with for figuring out how many digits to show is basically:
- find the minimum distance from data points to bin edges from either side - this is just two numbers for the whole distribution
- for two consecutive bins, find the largest digit that changes in both the left and right gaps (with a little shifting to account for rounding errors)

This algorithm is a little complicated to state, but it's quick to calculate and it's robust as far as I can see, on numeric, category, and gregorian date axes (though even in that case there is a weird [edge case](https://github.com/plotly/plotly.js/pull/2113/commits/6cb4e7130e4a5a1de7e5819fc6eb7c487c7997a4#diff-8c89958f1286cac36ab2325921671724R819) that confuses it). However, the larger variability of month and year lengths in world calendars causes it to sometimes think a digit doesn't change when in the actual data it does, and increasing the tolerance to catch this breaks some much more common cases.

I suspect we could fix this by looking at *all* bin edges independent of each other, but that's far more computationally intensive, and has some other potential weird pitfalls. I don't think providing a bit of extra precision in the label for a few world calendar cases justifies all that extra complexity.

File: src/components/fx/hover.js (hunk 2 above)
Comment (alexcjohnson) [reply src/components/fx/hover.js]: yeah, better to use `d.(x|x)Label` -> https://github.com/plotly/plotly.js/pull/2113/commits/51ad8c239653ba0bb69387cefd791409fa6208db

File: src/plots/cartesian/axes.js [hunk 6]
```diff
@@ -1221,7 +1221,7 @@ axes.hoverLabelText = function(ax, val) {
     var tx = axes.tickText(ax, ax.c2l(logOffScale ? -val : val), 'hover').text;
 
     if(logOffScale) {
-        return val === 0 ? '0' : '-' + tx;
+        return val === 0 ? '0' : MINUS_SIGN + tx;
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js]: been around for a while, using regular dash instead of unicode minus. tested in https://github.com/plotly/plotly.js/pull/2113/commits/3aa03f7911fe6988fafcc71c1451b9c9f17bb76f#diff-846cf5b534aa0d22bdd1da2b43ac3cbaR517

File: test/jasmine/tests/hover_label_test.js [hunk 7]
```diff
@@ -501,6 +501,128 @@ describe('hover info', function() {
...
+            }], {
+                width: 500,
+                height: 400,
+                margin: {l: 0, t: 0, r: 0, b: 0}
+            })
+            .then(function() {
+                _hover(gd, 250, 200);
+                assertHoverLabelContent({
+                    nums: '3',
+                    axis: '3.3'
```
Comment (alexcjohnson) [inline test/jasmine/tests/hover_label_test.js]: ⚠️  another potentially controversial decision: whenever *every* bin contains a single unique value, label exactly that value, not a range, even if those single values are not at the same places within the bins. But as soon as you have any bin with two distinct values in it, we use the range logic.

Note: for histogram2d I applied this independently to each axis, not to each brick. So if one x brick has only the point (2, 3) and one above it has only the point (2.1, 4), this will make the x bins show ranges, because *that one x bin* has multiple values, even though *each brick* has unique values in it.

Comment (etpinard) [reply test/jasmine/tests/hover_label_test.js]: Nice touch here. I was worried that category histograms would get e.g, [`apples - lemons`](http://rickyreusser.com/plotly-mock-viewer/#hist2d_summed) bin hover labels. :ok_hand:

File: test/jasmine/tests/hover_label_test.js [hunk 8]
```diff
@@ -501,6 +501,128 @@ describe('hover info', function() {
...
+                xbins: {start: -0.5, end: 8.5, size: 3},
+                type: 'histogram'
+            }], {
+                width: 500,
+                height: 400,
+                margin: {l: 0, t: 0, r: 0, b: 0}
+            })
+            .then(function() {
+                _hover(gd, 250, 200);
+                assertHoverLabelContent({
```
Comment (etpinard) [inline test/jasmine/tests/hover_label_test.js]: Would mind checking that `plotly_hover` does emit the correct event data?

File: src/traces/histogram/calc.js (hunk 3 above)
Comment (etpinard) [reply src/traces/histogram/calc.js]: > so that the hover label at least gets the description "all the data prior to X"

Yeah. I think something along those lines would be useful down the road. 

> You could say it should be included and again should be "all the data prior to X" but then it would be meaningless to select a single bar and not all the bars before it. Alternatively pts could contain "all the data that was added in this bar" 

Sounds to me like both _all-data-prior-to-X_ and _data-added-in-this-bar_ point lists could be useful, so we'll might have to emit both in the future. Leaving this discussion for another time is :ok_hand: with me. I doubt that most plotly.js users even know about cumulative histograms.

File: test/jasmine/tests/histogram_test.js (hunk 5 above)
Comment (alexcjohnson) [reply test/jasmine/tests/histogram_test.js]: The issue with dates is that the actual time increments - the "digits" we're looking to see if they have changed - have different sizes in terms of milliseconds. The *bin sizes* being different should not be a problem, though people could do things to thwart the algorithm, like having the first two bin edges on fairly round values but some subsequent edges at not-so-round values. The idea that two edges are enough to determine the biggest digit that always changes comes from the observation that if *one* bin is extra-round, its neighbor will not be. So to do this robustly with nonuniform bins, we could either:
- Do this same algorithm on *all* bin edges. That's different from looking at all bin edges independently as I suggested above for world calendars, though that would solve this problem too - what I'm suggesting here is to calculate just one left-edge-gap and one right-edge-gap as we do now, but then test those gaps against all bin edges instead of the first two.
- As a last resort, make a new attribute for folks to set their own rounding.

One common case mentioned in #360 is equally-sized bins on a log axis. There, the whole assumption of a single digit to round to is wrong, and we'll have to do something completely different or just give up on the idea of disambiguating the edges, and let the regular axis-formatting-for-hover system handle it unchanged.

File: src/traces/histogram/calc.js (hunk 3 above)
Comment (alexcjohnson) [reply src/traces/histogram/calc.js]: Great, I'll make a new issue  to discuss how this should work for CDFs.

File: test/jasmine/tests/hover_label_test.js (hunk 7 above)
Comment (alexcjohnson) [reply test/jasmine/tests/hover_label_test.js]: That *can* happen - but only if the bins *actually* contain apples and lemons.
```js
Plotly.newPlot(gd, [{
  x: ['apples','apples','apples','lemons','lemons','peaches','peaches','pears'],
  type: 'histogram',
  xbins: {start: -0.5, end: 3.5, size: 2}
}],{
  width: 400,
  height: 400
})
```

screen shot 2017-10-24 at 11 43 19 am

It's possible to put as many categories as you want into one bin, and as with numbers we'll only show the first and last ones in the label. Never happens automatically since it's in general confusing, but if you want to do it we won't stop you!

Comment (etpinard) [reply test/jasmine/tests/hover_label_test.js]: > but if you want to do it we won't stop you!

Cool. Would you mind adding a test case for this situation?

File: test/jasmine/tests/hover_label_test.js (hunk 8 above)
Comment (alexcjohnson) [reply test/jasmine/tests/hover_label_test.js]: Good catch -> tested in https://github.com/plotly/plotly.js/pull/2113/commits/2696c1c9ded8fb57c61ef103f08c374dda2a4f53

File: test/jasmine/tests/hover_label_test.js (hunk 7 above)
Comment (alexcjohnson) [reply test/jasmine/tests/hover_label_test.js]: > Cool. Would you mind adding a test case for this situation?

https://github.com/plotly/plotly.js/pull/2113/commits/0331a16845be937de4cedac301979a0139249fa1

**Comment by etpinard** [comment]
Nicely done :dancer:
